react-js/react-part-2.md

```diff
@@ -35,19 +33,71 @@ In introduction to [React](/react.md) we discussed about how React makes website
 
 ### Guided Practice
 
-1. Define state.
-  - >	``` state = {} ```
-2. Initialize state
+- Define state.
+  ``` 
+  state = {} 
+  ```
+- Initialize state
   - State in class component can be initialized in two ways
-    - Directly inside the 
+    - Directly inside the class component, outside all methods.
+      ``` 
+      state = {
+        message: "Hello world"
+      }
+      ```
+
+    - Inside class constructor
+      ```
+      constructor(){
+        this.state = {
+          message: "Hello world"
+          }
+      }
+      ```
 
-3. State should not be mutated directly i.e. a copy of state should be created and then updated. (already covered in lesson)
+- State should not be mutated directly i.e. a copy of state should be created and then updated. (already covered in lesson)
     - Incorrect way -
-      - > ```this.state.list.push(value)```
+      ```
+      this.state.list.push(value)
+      ```
     - Correct way -
-      - > ``` this.setState( { ...this.state , list: [ ...this.state.list, value ] } )```
-    
-#### Now , let us create a simple todo list.
+      ``` 
+	  this.setState({
+		  ...this.state,
+		  list: [ ...this.state.list, value ]
+	  })
+	  ```
+- You must have understand till now that we should not update state directly, but what special thing *setState* function does to the state, that the component re-renders automatically. Well, in introduction to React, we have learnt that React follow a special path called the React cycle. So, when state is updated using setState, React goes into **shouldComponentUpdate** and **componentWillUpdate** method (in case of state update and an additional method in case of change of props i.e. **componentWillReceiveProps**) of the React cycle and compare the changes, changes are done accordingly and those changes are reflected in the UI. 
+  - React provides a declarative API so that you don’t have to worry about exactly what changes on every update. This process is called reconciliation. 
+  - The reconciliation process is the way React updates the DOM, by making changes to the component based on the change in state.
+  - When the request to setState() is triggered, React creates a new tree containing the reactive elements in the component (along with the updated state).
+  - React knows which changes to implement and will only update the parts of the DOM where necessary.
```

Comment:
This is unfortunately too much detail at this stage of learning. If you want to link to a detailed explanation in the Supplemental Materials, feel free. Otherwise, all you have to mention is that calling `setState` will cause a re-render.